go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,168 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	ipam "k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+
+	"reflect"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"k8s.io/klog"
+)
+
+// logicalSwitchInfo contains information corresponding to the node. It holds the
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+// logicalSwitchInfo is used to save switch specific information such as subnets,
+// allocated IPs per switch etc.
+type logicalSwitchInfo struct {
+	hostSubnets []*net.IPNet
+	ipams       []ipam.Interface
+}
+
+type logicalSwitchCache struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchCache which holds logicalSwitch information
+	sync.RWMutex
+}
+
+type ipamFactoryFunc func(*net.IPNet) ipam.Interface
+
+func newIPAMAllocator(cidr *net.IPNet) ipam.Interface {
+	return ipam.NewCIDRRange(cidr)
+}
+
+func (cache *logicalSwitchCache) Add(nodeName string, hostSubnets []*net.IPNet, ipamFunc ipamFactoryFunc) error {
```

Comment:
@abhat I'm not a huge fan of the pattern here for Add(). Can we just pass the factory function in when creating the logicalSwitchCache object, and not on Add()?

Reply:
We can. Although that means the `factoryfunc` will have to be part of the object which seemed un-necessary and prevents us from using different IPAMs for different nodes (fwiw).

Reply:
@abhat why would we use a different IPAM for different nodes?

Reply:
If we don't anticipate a need for that, I will change to let the factory func be passed during creation of the lsCache.